components/service/glean/src/main/java/mozilla/components/service/glean/StringListMetricType.kt

```diff
@@ -0,0 +1,105 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean
+
+import kotlinx.coroutines.launch
+import mozilla.components.service.glean.storages.StringListsStorageEngine
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This implements the developer facing API for recording string list metrics.
+ *
+ * Instances of this class type are automatically generated by the parsers at build time,
+ * allowing developers to record values that were previously registered in the metrics.yaml file.
+ *
+ * The string list API exposes the [add] and [set] methods, which take care of validating the input
+ * data and making sure that limits are enforced.
+ */
+data class StringListMetricType(
+    override val disabled: Boolean,
+    override val category: String,
+    override val lifetime: Lifetime,
+    override val name: String,
+    override val sendInPings: List<String>
+) : CommonMetricData {
+
+    override val defaultStorageDestinations: List<String> = listOf("metrics")
+
+    private val logger = Logger("glean/StringListMetricType")
+
+    companion object {
+        // Maximum length of any passed value string, in UTF8 byte sequence length.
+        const val MAX_STRING_LENGTH = 50
+    }
+
+    /**
+     * Appends a string value to one or more string list metric stores.  If the string exceeds the
+     * maximum string length, defined as [MAX_STRING_LENGTH], it will be truncated.
+     *
+     * If adding the string to the lists would exceed the maximum value defined as
+     * [StringListsStorageEngine.MAX_LIST_LENGTH_VALUE], then the storage engine will drop the new
+     * value and it will not be added to the list.
+     *
+     * @param value This is a user defined string value. The maximum length of
+     * this string is [MAX_STRING_LENGTH].
+     */
+    fun add(value: String) {
+        // TODO report errors through other special metrics handled by the SDK. See bug 1499761.
+
+        if (!shouldRecord(logger)) {
+            return
+        }
+
+        val truncatedValue = value.let {
+            if (it.length > MAX_STRING_LENGTH) {
+                logger.warn("$category.$name - string too " +
+                        "long ${it.length} > $MAX_STRING_LENGTH")
+                return@let it.substring(0, MAX_STRING_LENGTH)
+            }
+            it
+        }
+
+        Dispatchers.API.launch {
+            // Delegate storing the string to the storage engine.
+            StringListsStorageEngine.add(
+                    this@StringListMetricType,
+                    value = listOf(truncatedValue)
```

Comment:
Since `add` is only supporting adding a single string, what about changing `StringListsStorageEngine.add` to accept a single string rather than a list? This way we would get rid of `listOf`.

Reply:
I had originally implemented it this way thinking we may want an 'addAll()' function and this allowed for reuse of the function.  If we know we don't want an addAll() function then it makes sense to change it to be singular.

Reply:
I see. This is a good point. @mdboom , thoughts? Should we have an API function to add a list of strings in addition to a single string?

Reply:
I'm not sure -- I don't know if I understand the use cases well enough.  For the one use case I'm aware of (accessibility services in the baseline ping), it would probably be convenient to have an addAll, so I guess I'm a weak :+1: on having that.

Reply:
@mdboom mh, would this usecase be covered by the `set` method we already expose?

Reply:
Currently, `set()` replaces the list and `add()` appends to it.  So they would have to get the list from `getSnapshot()` and `addAll()` to it and then pass it back to `set()`

Reply:
@travis79 yes, but I think that for the accessibility use-case values are just inserted once, at startup, all at the same time: exactly what `set` does. Anyway, let's keep the API surface small unless we find a strong use-case to power new API methods. We can consider adding an `addAll` method later, if needed.